### DISCUSSION TITLE
Is it possible to run a script or wakeup macOS after macOS in sleeping mode?

### DISCUSSION THREAD

**Discussion by bsdshell** [Hammerspoon/hammerspoon#3053] [General]
I'm wondering whether it is possible to wake up macOS after macOS in sleeping mode ?

**Comment by luckman212** [comment]
Not quite sure what you're asking - but if you want to run a script after your Mac wakes up, check out [sleepwatcher](https://www.bernhard-baehr.de/).

**Reply by cmsj** [reply]
See also http://www.hammerspoon.org/docs/hs.caffeinate.watcher.html 🙂

**Comment by cmsj** [comment]
> I'm wondering whether it is possible to wake up macOS after macOS in sleeping mode ?

Are you thinking of something like WakeOnLAN?

I think we need a bit more detail about what you're wanting to do, would be helpful.

**Reply by bsdshell** [reply]
In macOS, If you are running cron job, you cron job will stop running if the macOS is in sleep mode.

For example:
If I want to run my script or some cron job at 3am at night, it seems to be not possible after macOS is in sleep mode. (correct me If I'm wrong)

My macOS will be in sleep mode around 1:am since I sleep around 1am.

My question is whether your API can wake up my macOS at 3am so that I can run some script or run some cron job.

**Reply by luckman212** [reply]
You can use the built in Energy Saver control panel settings to do this. Look under Schedule. You can set the Mac to wake itself every day at 3am to run your scripts, and then just have your script (or Hammerspoon with `hs.caffeinate.systemSleep()`) put the Mac back to sleep when it's finished.

**Reply by bsdshell** [reply]
Thanks, I will check it out.

**Comment by cmsj** [comment]
I think it's possible to programatically schedule those wakeups - I note that Carbon Copy Cloner does it at least. Maybe we should figure out how that works and put it in to `hs.caffeinate`

**Reply by luckman212** [reply]
Yes, seems like you're right @cmsj 

https://developer.apple.com/documentation/iokit/1557076-iopmschedulepowerevent

**Reply by cmsj** [reply]
Looks like it needs root access to call though 😕

**Reply by luckman212** [reply]
Hmm, guess we'll need a "Hammerspoon Helper" LaunchDaemon/PrivilegedHelperTool now 😄
(That's how CCC does it...)

![image](https://user-images.githubusercontent.com/1992842/150345731-7164e3a9-8a85-4aaf-996a-f073009e91b4.png)
